### PR TITLE
Support Apple MLX as a separate backend to Numpy

### PR DESCRIPTION
**Pull request by workofart**
## Summary

This PR refactors the library toward a backend-selectable runtime, with MLX support on macOS, CuPy support on CUDA hosts, and NumPy support still available through the shared backend layer.

The motivation is to speed up development on a Mac with Metal acceleration, as opposed to finding a Nvidia GPU server.

## Benchmark Snapshot

GPT-2 benchmark audit against NumPy `main`, this MLX branch, and PyTorch/MPS shows the MLX migration is a real step up from the NumPy baseline, while PyTorch/MPS is still faster once warm. 

Hyperparameters: dropout_prob=0.0, adam_epsilon=1e-7, batch_size=16, seq_len=96, steps_per_epoch=16, total_epochs=3.

  | Backend | Train time | Throughput | Steady-state step | Final train loss | Final val loss |
  | --- | --- | --- | --- | --- | --- |
  | NumPy backend | 79.1140s | 931.92 tok/s | 1.5729s | 6.1618 | 6.0967 |
  | MLX backend | 7.6220s | 9673.50 tok/s | 0.1423s | 6.1692 | 6.1025 |
  | PyTorch / MPS | 4.4620s | 16522.40 tok/s | 0.0832s | 6.1686 | 6.1017 |

Speed comparisons:
- MLX vs NumPy: 10.38x faster end-to-end, 11.05x faster steady-state.
- PyTorch / MPS vs MLX: 1.71x faster end-to-end, 1.71x faster steady-state.
- Benchmark caveat: the 3-way comparison used a transient external harness; the checked-in branch diff does not add that full benchmark harness.
- One of the biggest reasons for PyTorch to be faster than MLX, despite both running on Apple Silicon GPU, is the optimized attention kernels in PyTorch (e.g. FastAttention). This gap will be closed later when we implement those optimizations step-by-step as an education process

## What Changed

- Added `autograd/backend.py` to centralize backend-specific behavior across MLX, CuPy, and NumPy and provide shared helpers used by the tensor, functional, nn, optim, and tooling modules.
- Migrated the core tensor/autograd stack and model code onto the shared backend surface, including updates across `autograd/tensor.py`, `autograd/functional.py`, `autograd/nn.py`, `autograd/optim.py`, `autograd/init.py`, and the training/data utilities, with MLX now supported as a first-class backend.
- Fixed `cross_entropy(..., label_smoothing=...)` parity and added explicit PyTorch comparison coverage for both 2D and 3D logits.
- Updated tests to use shared backend-aware helpers and added focused backend regression coverage in `test/autograd/test_backend.py`, including CuPy detection behavior.
- Switched project setup and CI to the backend-selectable runtime model: `uv.lock` / `uv sync`, backend-aware GitHub Actions coverage, bootstrap updates, optional CuPy extras, and README updates documenting backend auto-selection and overrides.
- Renamed `examples/gpt-2.py` to `examples/gpt_2.py` and refreshed examples/docs to match the MLX branch behavior.

## Testing
  - uv run ruff check
  - uv run ruff format . --check
  - uv run pyright
  - AUTOGRAD_BACKEND=mlx uv run pytest
  - AUTOGRAD_BACKEND=numpy uv run pytest

### CODE REVIEW

File: autograd/backend.py
Comment (workofart) [inline autograd/backend.py:1]: This is the main addition to switch between different backends. 
- There are quite a few functions that differ in their APIs between MLX and Numpy, so I had to wrap them with the private compatability functions (e.g. `_scatter_add`)
- I tried to avoid creating another abstraction layer, so this file contains some if-else statements and the math is still preserved in the respective `tensor.py`, `functional.py` modules.
- The API is similar to previous Numpy's API (e.g. `xp.array` instead of `np.array`).
